src/Sulu/Bundle/PreviewBundle/Resources/js/views/Preview/Preview.js

```diff
@@ -0,0 +1,265 @@
+    @observable iframeRef: ?Object;
+    @observable token: ?string;
+    @observable started: boolean = false;
+
+    typeDisposer: () => void;
+    dataDisposer: () => void;
+
+    componentDidMount() {
+        if (previewConfigStore.mode === 'auto') {
+            this.startPreview();
+        }
+    }
+
+    @action startPreview = () => {
+        const {
+            resourceStore,
+            formStore,
+            router: {
+                attributes: {
+                    locale,
+                },
+            },
+        } = this.props;
+
+        const route = previewConfigStore.generateRoute('start', {
+            provider: resourceStore.resourceKey,
+            locale: locale,
+            id: resourceStore.id,
+        });
+        if (!route) {
+            return;
+        }
+
+        Requester.get(route).then((response) => {
+            this.setToken(response.token);
+        });
+
+        this.typeDisposer = autorun(() => {
+            if (formStore.loading || !this.iframeRef) {
+                return;
+            }
+
+            // $FlowFixMe
+            this.updateContext(formStore.type.get());
+        });
+
+        this.dataDisposer = autorun(() => {
+            if (resourceStore.loading || !this.iframeRef) {
+                return;
+            }
+
+            this.updatePreview(toJS(resourceStore.data));
+        });
+
+        this.started = true;
+    };
+
+    updatePreview = debounce((data) => {
+        const {
+            router: {
+                attributes: {
+                    locale,
+                    webspace,
+                },
+            },
+        } = this.props;
+
+        const route = previewConfigStore.generateRoute('update', {
+            locale: locale,
+            webspace: webspace,
+            token: this.token,
+        });
+        if (!route) {
+            return;
+        }
+
+        Requester.post(route, {data: data}).then((response) => {
+            const document = this.getPreviewDocument();
+            if (!document) {
+                return;
+            }
+
+            document.open();
+            document.write(response.content);
+            document.close();
+        });
+    }, previewConfigStore.debounceDelay);
+
+    updateContext = debounce((type) => {
+        const {
+            router: {
+                attributes: {
+                    webspace,
+                },
+            },
+        } = this.props;
+
+        const route = previewConfigStore.generateRoute('update-context', {
+            webspace: webspace,
+            token: this.token,
+        });
+        if (!route) {
+            return;
+        }
+
+        Requester.post(route, {context: {template: type}}).then((response) => {
+            const document = this.getPreviewDocument();
+            if (!document) {
+                return;
+            }
+
+            document.open();
+            document.write(response.content);
+            document.close();
```

Comment:
That's not the first time I read that code :wink: I think it would really make sense to extract that in some way.